File: zcat/codesin.py
#!/usr/bin/pypy3

import sys

class LZW:
    def __init__(self, maxbits):
        self.maxbits = maxbits
        self.oldcode = self.finchar = 0
        self.dict = list()
    def code(self, incode):
        c = incode
        assert c <= len(self.dict) + 256
        if c == 256:
            self.dict.clear()
            return bytearray()
        stack = bytearray()
        if c == len(self.dict) + 256:
            stack.append(self.finchar)
            c = self.oldcode
        while c >= 256:
            stack.append(self.dict[c - 256][1])
            c = self.dict[c - 256][0]
        self.finchar = c
        if len(self.dict) + 256 < 1 << self.maxbits:
            self.dict.append((self.oldcode, self.finchar))
        stack.append(self.finchar)
        stack.reverse()
        self.oldcode = incode
        return stack

if __name__ == "__main__":
    lzw = LZW(16)
    for line in sys.stdin:
        sys.stdout.buffer.write(lzw.code(int(line)))

File: zcat/codesin.cpp
#include <vector>
#include <iostream>
#include <cassert>

using std::ostream;
using std::string;
using std::cout;
using std::cin;

class PrintStack
{
    std::vector<char> _stack;
public:
    void push_back(char c) { _stack.push_back(c); }
    void print(ostream &os) { for (; _stack.size(); _stack.pop_back()) os.put(_stack.back()); }
};

class LZW
{
    const unsigned _maxbits;
    unsigned _oldcode = 0;
    char _finchar;
    std::vector<std::pair<unsigned, char>> _dict;
    ostream &_os; 
    PrintStack _stack;
public:
    LZW(unsigned maxbits, ostream &os) : _maxbits(maxbits), _os(os) { }

    inline void code(const unsigned in)
    {
        assert(in <= _dict.size() + 256);
        auto c = in;

        if (in == 256)
        {
            _dict.clear();
            return;
        }

        if (c == _dict.size() + 256)
        {
            _stack.push_back(_finchar);
            c = _oldcode;
        }

        for (; c >= 256U; c = _dict[c - 256].first)
        {
            _stack.push_back(_dict[c - 256].second);
        }

        _os.put(_finchar = c);

        if (_dict.size() + 256 < 1U << _maxbits)
        {
            _dict.push_back(std::pair<unsigned, char>(_oldcode, _finchar));
        }

        _oldcode = in;
        _stack.print(_os);
    }
};

int main()
{
    string s;
    LZW lzw(16, cout);

    while (std::getline(cin, s))
        lzw.code(std::stoi(s));

    return 0;
}


